source/Halibut/Transport/Protocol/MessageExchangeProtocol.cs

```diff
@@ -207,26 +208,36 @@ async Task ProcessSubscriberAsync(IPendingRequestQueue pendingRequests, Cancella
             }
         }
         
-        async Task<bool> ProcessReceiverInternalAsync(IPendingRequestQueue pendingRequests, RequestMessage nextRequest, CancellationToken cancellationToken)
+        async Task<bool> ProcessReceiverInternalAsync(IPendingRequestQueue pendingRequests, RequestMessageWithCancellationToken nextRequest, CancellationToken cancellationToken)
         {
             try
             {
                 if (nextRequest != null)
                 {
-                    var response = await SendAndReceiveRequest(nextRequest, cancellationToken);
-                    await pendingRequests.ApplyResponse(response, nextRequest.Destination);
+                    using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(nextRequest.CancellationToken, cancellationToken);
```

Comment:
The cancellation token on the Request is now observed so we can cancel the thread that is dequeuing the pending request